text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+
+        Ok(((), mem::size_of::<T>()))
+    }
+
+    unsafe fn grow_zeroed(
+        &self,
+        _handle: Self::Handle,
+        old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError> {
+        debug_assert!(
+            new_layout.size() >= old_layout.size(),
+            "{new_layout:?} must have a greater size than {old_layout:?}"
+        );
+
+        Self::validate_layout(new_layout)?;
+
+        let pointer = self.0.get() as *mut u8;
+
+        //  Safety:
+        //  -   Both starting and resulting pointers are in bounds of the same allocated objects as `old_layout` fits
+        //      `pointer`, as per the pre-conditions of `grow_zeroed`.
+        //  -   The offset does not overflow `isize` as `old_layout.size()` does not.
+        let pointer = unsafe { pointer.add(old_layout.size()) };
+
+        //  Safety:
+        //  -   `pointer` is valid, since `self` is valid.
+        //  -   `pointer` points to at an area of at least `mem::size_of::<T>() - old_layout.size()`.
+        //  -   Access to the next `mem::size_of::<T>() - old_layout.size()` bytes is exclusive.
+        unsafe { ptr::write_bytes(pointer, 0, mem::size_of::<T>() - old_layout.size()) };
+
+        Ok(((), mem::size_of::<T>()))
+    }
+}
+
+//  Safety:
+//  -   `self.resolve(handle)` always returns the same address, as long as `self` doesn't move.
+unsafe impl<T> StableStore for InlineSingleStore<T> {}
+
+impl<T> fmt::Debug for InlineSingleStore<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        let layout = Layout::new::<T>();
+
+        f.debug_struct("InlineSingleStore")
+            .field("size", &layout.size())
+            .field("align", &layout.align())
+            .finish()
+    }
+}
+
+impl<T> InlineSingleStore<T> {
+    fn validate_layout(layout: Layout) -> Result<(), AllocError> {
+        let own = Layout::new::<T>();
+
+        if layout.align() <= own.align() && layout.size() <= own.size() {
+            Ok(())
+        } else {
+            Err(AllocError)
+        }
+    }
+}
+```
+
+And that's it!
+
+I need not implement `MultipleStore`, and thus do not have to track allocations and deallocations. And I need not
+implement `PinningStore`, and thus do not have to ensure memory address stability across moves.
+
+More examples of `Store` can be found at https://github.com/matthieu-m/storage/tree/main/src/store, including an inline
+bump store.
+
+
+#   Reference-level explanation
+
+##  Overview
+
+This RFC introduces 4 new traits.
+
+The core of this RFC is the `Store` trait:
+
+```rust
+/// Allocates and deallocates handles to blocks of memory, which can be temporarily resolved to pointers to actually
+/// access said memory.
+pub unsafe trait Store {
+    /// Handle to a block of memory.
+    type Handle: Copy;
+
+    /// Returns a dangling handle, always invalid.
+    fn dangling() -> Self::Handle;
```

Comment:
Would it make sense to take a `Layout` or [`ptr::Alignment`](https://doc.rust-lang.org/std/ptr/struct.Alignment.html) in `dangling`, so that `Vec::new` can use dangling to get the 0-length slice pointer?

Reply:
That's a good question.

My intent was for dangling handles to always be invalid, and therefore to be UB to do anything with it, including resolving it. For example, I hoped that an index could use a giant value, which `resolve` would be able to detect was out of bounds in Debug.

Needless to say, this doesn't mesh well with `Vec::as_ptr`, which is then used for 0-length slices. It requires `Vec` to introduce a branch in `as_ptr` to choose between resolving or returning a dangling pointer, and that branch may not be well-optimized at all... it's all around sad.

This leaves two choices, as far as I can see:

1. Alter the invalidity of dangling handle, and state that they _must_ be resolvable, though the resolved pointer itself is then invalid. It would still be invalid to attempt to grow/shrink/deallocate a dangling handle, however.
2. Require `Vec` to use a 0-sized allocation, and implementers to handle those.

I believe (1) is clearly the superior alternative here, so I'll go ahead and amend the RFC in that direction in the next batch of fixes, and mention why in the alternatives section.

Reply:
So... I went ahead and tried having `dangling` take a `ptr::Alignment` argument, and there was an unforeseen circumstance: it became fallible.

I created a branch on the companion repository [which I invite you to look at](https://github.com/matthieu-m/storage/tree/experiment/dereferenceable-dangling-handle). The problem I face there is that inline stores cannot provide an allocation with an alignment greater than their own as allocations are always offsets within the store itself. In turn, this makes `Vec::new()` fallible, or in the companion repository branch it makes `LinkedList::new()` and `SkipList::new()` fallible.

This result, in turn, in making `dangling` fallible since one may (accidentally, I suppose) ask for an alignment greater than what the Store may be able to provide, even in the absence of actual allocation.

I am quite unsure of how to move forward on this.

Reply:
In https://github.com/rust-lang/rfcs/pull/3446#discussion_r1240498730 @CAD97 raised the possibility of an `allocate_dangling` method.

It would make the interface more complex, but we could have:

- `Store::invalid`: an invalid handle, it should never be resolved, deallocated, grown, nor shrunk.
- `Store::allocate_zero_sized`: a specialized allocation method for ZSTs. The handle is valid, and I would argue should be treated like any other handle: it may be resolved, deallocated, grown, etc... (it can't be shrunk). Multiple independent ZST handles may resolve to the same memory address. Unlike CAD97's proposal I'd argue it must be deallocated ultimately.

This would help making `LinkedList::new` and `SkipList::new` infallible again, as they could use `Store::invalid`. It would not, unfortunately, help `Vec::new` be infallible as `Store::allocate_zero_sized` would remain fallible due to the alignment argument.

Reply:
> Unlike CAD97's proposal I'd argue it must be deallocated ultimately.

This is a problem for implementations. It also leads to the situations we see where both sides of the allocation interface end up checking for ZSTs and having to special case them (See my links to the ZST allocation issues with the current API elsewhere in the thread). While that's less detrimental for Stores (Store is probably not going to be frequently, if at all, used via virtual dispatch), it's still a bad sign IMO.

Reply:
> an unforeseen circumstance: `dangling` became fallible.

Fallibility wasn't unforseen on my end; my `allocate_dangling` did return `Result` to allow failure to allocate if the alignment request cannot be fulfilled. It's fine if `Vec::new` can fail and call `handle_alloc_error`, imho; the entire collection interface is built around assuming that allocation is essentially infallible. A `try_new_in` (or whatever solution is used for fallible allocation in collections) would be necessary instead to avoid termination on exhausting the storage.

The ability to skip deallocating an `allocate_dangling` handle is key to it being any amount of realistically useful. Without that guarantee, it's unambiguously better to just never use the allocator to create zero-sized allocations.

Note that my spitball still allowed zero sized allocations in the standard `allocate` path, requiring deallocation to prevent leaks, with the expectation that this allocation would be permitted to use the exact same path as non-zero-sized allocation and be wasteful for zero-sized allocation which can conjure a fresh allocation at any nonzero aligned address.

> `Store` is probably not going to be frequently, if at all, used via virtual dispatch

It'd be used dynamically essentially exactly as much as `Allocator` would, since it's replacing `Allocator` as the storage generic axis for collections. FWIW, I fully expect `dyn GlobalAlloc` will always remain the interface used for `#[global_allocator]`, and changing it to be `dyn Allocator` to be unnecessary and of no real benefit. It makes sense that the API tradeoff calculus would be different for the globally available static allocation interface. (Though automatically bridging from `Allocator` to `GlobalAlloc` is both reasonable and desirable.)

Reply:
> It's fine if `Vec::new` can fail and call `handle_alloc_error`, imho; the entire collection interface is built around assuming that allocation is essentially infallible.

I am coming around to thinking this may be fine.

By defining `type InlineVec<T, const N: usize> = Vec<T, InlineSingleStore<[T; N]>>;` the type is guaranteed to have an infallible `dangling`, and thus users of the abstraction will never see a panicking `new`.

Performance-wise, this would rely on `dangling` being inlined in `new`, so the optimizer can see the error path is never taken and optimize out any check/call, but that can be tamed.

Reply:
> It's fine if Vec::new can fail and call handle_alloc_error, imho; the entire collection interface is built around assuming that allocation is essentially infallible

I'm not sure this is fine, it seems quite unfortunate to me. Note that `Vec::new` is currently const. Obviously that only applies to the global allocator (or whatever default store is used for Vec), but it seems like quite a drawback to prevent things like `Store`-based smallvec/arrayvec/etc from having `const fn new`s.

Reply:
> but it seems like quite a drawback to prevent things like `Store`-based smallvec/arrayvec/etc from having `const fn new`s.

Note that `handle_alloc_error` is (unstably) `const`, so it won't prevent `InlineVec::new` from being `const`.

_On the other hand, Rust doesn't support marking individual trait functions `const` for now, so we may need to lobby the compiler team or otherwise find a clever work-around for `Store::dangling`._

Reply:
@scottmcm For your viewing pleasure, let me present [`StoreVec`](https://github.com/matthieu-m/storage/blob/72c9818cf64e859ed33b23353381a350df03cb43/src/collection/store_vec.rs#L495-L525). As a bonus, built entirely in a `const` context, because I could.

There's only one glitch preventing `StoreVec::new` to be `const` in general: `Default` is not a `#[const_trait]` yet, sadly.